README.md
# 👋🏻 Welcome to FORM-ONE

## Why?

form-one is built onto the react-hook-form library (which is an amazing library itself 🦄).
form-one it is an abstraction of the react-hook-form library.
form-one let you focusing just on the form api interface, which is really simple.
## How?
The *ui* is made with ❤️ using the antd library and the *core* is built abstracting the react-hook-form library.

## How to use form-one?
You can use [form-one](https://github.com/Lincerossa/form-one) in React apps by just installing it
```
  npm i form-one
  import FormOne from 'form-one'

```

## Why it is so simple?
These are the available props:
```
  <FormOne
    onChange
    onSubmit
    inputs
    layout
    validationSchema
    initialValues
  />

```

Let's have a look about how it works in a real demo, by visiting it's [playground](https://codesandbox.io/s/form-one-playground-g9uwg?file=/src/App.js) in Codesandbox or its [storybook](https://form-one.netlify.app/)


CHANGELOG.md
# [1.5.0](https://github.com/Lincerossa/form-one/compare/v1.4.2...v1.5.0) (2021-03-01)


### Features

* lib update ([94b406e](https://github.com/Lincerossa/form-one/commit/94b406e37c955ab051f2ab05ac69a5d915f4cb49))

## [1.4.2](https://github.com/Lincerossa/form-one/compare/v1.4.1...v1.4.2) (2021-03-01)


### Bug Fixes

* 🐞 RenderPreview has to be removed from nested repeaters ([669d5fc](https://github.com/Lincerossa/form-one/commit/669d5fc3a9d38a6c7ca8c200b8c9d4646b5c10c5))

## [1.4.1](https://github.com/Lincerossa/form-one/compare/v1.4.0...v1.4.1) (2021-03-01)


### Bug Fixes

* removed not existing import ([be7098d](https://github.com/Lincerossa/form-one/commit/be7098d8cb081beb83721853ea741f16963ab28c))

# [1.4.0](https://github.com/Lincerossa/form-one/compare/v1.3.1...v1.4.0) (2021-03-01)


### Features

* 🚀 now the Repeater has a brand new functionality. it can render a preview or an edit mode. ([001eab7](https://github.com/Lincerossa/form-one/commit/001eab7b9ad4f16b7afde2c8126080134d9cdbdd))

## [1.3.1](https://github.com/Lincerossa/form-one/compare/v1.3.0...v1.3.1) (2021-02-19)


### Bug Fixes

* minor readme update ([4ac3b7a](https://github.com/Lincerossa/form-one/commit/4ac3b7af3005af73b2bb4d9d9d46295bad4279f8))

# [1.3.0](https://github.com/Lincerossa/form-one/compare/v1.2.0...v1.3.0) (2021-02-18)


### Features

* 📖 docs update ([41f18c4](https://github.com/Lincerossa/form-one/commit/41f18c49b8e32e2468f85280c9c13b84207ce7c9))

# [1.2.0](https://github.com/Lincerossa/form-one/compare/v1.1.0...v1.2.0) (2021-02-18)


### Features

* 🐞 the transpile process was bugged, now it should properly works ([165588e](https://github.com/Lincerossa/form-one/commit/165588e7d30c937c03a7ebb1b7ce7365240e7350))

# [1.1.0](https://github.com/Lincerossa/form-one/compare/v1.0.0...v1.1.0) (2021-02-18)


### Features

* force new release (by writing the author name in the packge.json) ([eee4e8b](https://github.com/Lincerossa/form-one/commit/eee4e8bf1cf330f2153db03af7f0ae381a249739))
* force release ([e43e9d7](https://github.com/Lincerossa/form-one/commit/e43e9d740cd7647cd45e094c313a0d829ced29aa))


src/Form/stories.js
import React, { useState } from 'react'
import * as Yup from 'yup'
import Form from './index'

const getRandomInputItems = (length, label) => Array.from({ length }, (e, i) => ({
  label: `${label}${i + 1}`,
  value: Math.random(),
}))

const inputs = [
  {
    type: 'divider',
    name: 'divider',
    orientation: 'left',
    label: 'My main form',
  },
  {
    label: 'First Name',
    name: 'firstName',
    type: 'input',
    defaultValue: 'default value',
    placeholder: 'Egs',
  },
  {
    label: 'Custom Input',
    name: 'customInput',
    type: 'custom',
    CustomRender: ({ name }) => (
      <small>
        Beautiful input
        {' '}
        <strong>{name}</strong>
      </small>
    ),
  },
  {
    label: 'My birthday Name',
    name: 'myBirthday',
    type: 'datepicker',
  },
  {
    label: 'Age',
    name: 'myAge',
    type: 'number',
  },
  {
    label: 'radio example',
    name: 'radio1',
    type: 'radio',
    items: getRandomInputItems(4, 'radio'),
  },

  {
    label: 'sliderone',
    name: 'mySlider',
    type: 'slider',
    min: 1,
    max: 20,
  },
  {
    label: 'email',
    name: 'email',
    type: 'input',
    placeholder: 'test@email.com',
  },
  {
    label: 'Text area',
    name: 'myTextArea',
    type: 'textArea',
    placeholder: 'test@email.com',
  },
  {
    label: 'Tree',
    name: 'myTree',
    type: 'tree',
    items: [
      {
        title: 'parent 1',
        key: '0-0',
        children: [
          {
            title: 'parent 1-0',
            key: '0-0-0',
            children: [
              {
                title: 'leaf',
                key: '0-0-0-0',
                disableCheckbox: true,
              },
              {
                title: 'leaf',
                key: '0-0-0-1',
                children: [
                  {
                    title: 'leaf inner',
                    key: '0-0-0-0-1',
                    disableCheckbox: true,
                  },
                  {
                    title: 'leaffino',
                    key: '0-0-0-1-2',
                  },
                ],
              },
            ],
          },
          {
            title: 'parent 1-1',
            key: '0-0-1',
            children: [{ title: <span style={{ color: '#1890ff' }}>sss</span>, key: '0-0-1-0' }],
          },
        ],
      },
    ],
    placeholder: 'team eglobe',
  },
  {
    label: 'Checkbox input',
    name: 'myCheckbox',
    type: 'checkbox',
    items: getRandomInputItems(4, 'checkbox'),

    condition: ({ watch }) => {
      const myAge = watch('myAge')
      return myAge > 18
    },
  },
  {
    type: 'repeater',
    label: 'Repeater input',
    name: 'myRepeater',

    RenderPreview: ({ watch, name }) => {
      const data = watch(name)
      return (
        <div>
          <div>
            <strong>number:</strong>
            {data?.myNumber}
          </div>
        </div>
      )
    },
    items: [
      {
        type: 'number',
        defaultValue: 2,
        label: 'Choosen number in repeater',
        name: 'myNumber',
      },
      {
        type: 'select',
        label: 'select',
        defaultValue: 1,
        name: 'mySelect',
        items: [
          {
            value: 0,
            label: 'LABEL 1',
          },
          {
            value: 1,
            label: 'Label 2',
          },
        ],
      },
      {
        type: 'repeater',
        label: 'custom repeater inner',
        name: 'myRepeaterInner',
        items: [
          {
            type: 'number',
            label: 'Choosen number in repeater',
            name: 'myNumberInner',
            defaultValue: 4,
          },
          {
            type: 'select',
            label: 'select',
            name: 'mySelectInner',
            items: [
              {
                value: 0,
                label: 'LABEL 1',
              },
              {
                value: 1,
                label: 'Label 2',
              },
            ],
            condition: ({ watch, repeaterName }) => {
              const v = watch(repeaterName)
              return Number(v.myNumberInner) === 666
            },
          },
          {
            type: 'repeater',
            label: 'custom repeater inner',
            name: 'myRepeaterInnerInner',
            items: [
              {
                type: 'number',
                label: 'Choosen number in repeater',
                name: 'myNumberInnerInner',
                defaultValue: 422,
              },
              {
                type: 'select',
                label: 'select',
                name: 'mySelectInnerInner',
                items: [
                  {
                    value: 0,
                    label: 'LABEL 1',
                  },
                  {
                    value: 1,
                    label: 'Label 2',
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]

const validationSchema = Yup.object().shape({
  firstName: Yup.string().required('The Firstname is required').nullable(),
  myBirthday: Yup.string().required('The Birthday is required').nullable(),
  myAge: Yup.string().required('The Age is required').nullable(),
  myCheckbox: Yup.array().min(1).required('The Checkbox has to be checked somehow').nullable(),
  myRepeater: Yup.array().when('myAge', {
    is: (age) => age && age > 18,
    then: Yup.array()
      .of(
        Yup.object().shape({
          myNumber: Yup.string().required('Please, choose a number'),
          myRepeaterInner: Yup.array().of(
            Yup.object()
              .shape({
                myNumberInner: Yup.string().required('Please, choose a number'),
                myRepeaterInnerInner: Yup.array().of(
                  Yup.object()
                    .shape({
                      myNumberInnerInner: Yup.string().required('Please, choose a number'),
                    })
                    .required('Please, choose a number'),
                ),
              })
              .required('Please, choose a number'),
          ),
        }),
      )
      .required('The repeater has to be fulfilled correctly'),
  }),
})

const initialValues = {
  firstName: 'Marcello',
  myAge: 19,
  myRepeater: [
    {
      myNumber: 1,
      mySelect: 0,
      myRepeaterInner: [
        {
          myNumberInner: 2,
        },
      ],
    },
  ],
}

export default {
  title: 'Form',
  component: Form,
}

export const Default = () => (
  <Form
    layout="horizontal"
    inputs={inputs}
    onSubmit={(values) => console.log({ values })}
  />
)

export const Vertical = () => (
  <Form
    layout="vertical"
    inputs={inputs}
    initialValues={initialValues}
    validationSchema={validationSchema}
    onSubmit={(values) => console.log({ values })}
  />
)

export const Horizontal = () => (
  <Form
    layout="horizontal"
    inputs={inputs}
    initialValues={initialValues}
    validationSchema={validationSchema}
    onSubmit={(values) => console.log({ values })}
  />
)

export const onFormChange = () => {
  const inputsA = [...inputs]
  const inputsB = [{ type: 'TextArea', name: 'myBeautifulTextArea' }, ...inputs]
  const [currentInputs, setCurrentInputs] = useState(inputsA)

  function handleOnChange({ watch }) {
    const over18 = watch('myAge') > 18
    if (over18 && (currentInputs?.length !== inputsA.length)) setCurrentInputs(inputsA)
    if (!over18 && currentInputs?.length !== inputsB.length) setCurrentInputs(inputsB)
  }

  return (
    <Form
      layout="horizontal"
      inputs={currentInputs}
      onChange={handleOnChange}
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(values) => console.log({ values })}
    />
  )
}

export const StylingForm = () => (
  <Form
    layout="horizontal"
    inputs={inputs}
    style={{
      submitWrapper: {
        border: '3px dashed black',
        textAlign: 'center',
        padding: '1rem',
      },
      submit: {
        background: 'black',
        color: 'white',
      },
      form: {
        padding: '3rem',
        border: '2px dashed grey',
      },
    }}
    initialValues={initialValues}
    validationSchema={validationSchema}
    onSubmit={(values) => console.log({ values })}
  />
)
